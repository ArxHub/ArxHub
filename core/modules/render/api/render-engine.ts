import { VirtualFile } from '~/core/modules/vfs/api'
import { RenderOptions } from './render-options.ts'

export interface RenderEngine {
  render(file: VirtualFile, options: RenderOptions): Promise<string>
}
