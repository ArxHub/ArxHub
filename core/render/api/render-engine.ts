import { VirtualFile } from '../../api/vfs/mod.ts'
import { Renderer } from '../api/renderer.ts'
import { ErrorFactory } from '~/core/error'

export class RenderEngine {
  private readonly renderers: Renderer[]

  public constructor(renderers: Renderer[]) {
    this.renderers = renderers
  }

  public render(file: VirtualFile): Promise<string> {
    const renderer = this.getRenderer(file)
    return renderer.render(file)
  }

  private getRenderer(file: VirtualFile): Renderer {
    const renderer = this.renderers.find((renderer) => renderer.isSupported(file))
    if (renderer == null) throw ErrorFactory.FileRendererNotFound(file)
    return renderer
  }
}
