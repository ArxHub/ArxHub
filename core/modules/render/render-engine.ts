import { Plugin } from '../../plugin/mod.ts'
import { VirtualFile, VirtualFileSystem } from '~/core/modules/vfs/api'
import { CompositeFileLoader } from './composite-file-loader.ts'
import { Environment } from './environment.ts'
import { RenderOptions } from './api/render-options.ts'
import vento from '@third-party/vento'

export class RenderEngine {
  private readonly loader: CompositeFileLoader
  private readonly env: Environment

  constructor(vfs: VirtualFileSystem, plugins: Plugin<RenderEngine>[] = []) {
    this.loader = new CompositeFileLoader(vfs)
    this.env = vento({
      includes: this.loader,
      dataVarname: 'it',
      autoescape: false,
      autoDataVarname: true,
    })
    
    // TODO: use empty environment, instead of predefined vento
    // new Environment({
    //   loader: this.loader,
    //   dataVarname: 'it',
    //   autoescape: false,
    //   autoDataVarname: true,
    // })

    plugins.forEach((it) => it.apply(this))
  }

  async render(file: VirtualFile, options: RenderOptions): Promise<string> {
    const template = await this.loader.load(file)
    const data = { mode: options.mode, ...options.data, ...template.data }
    const { content } = await this.env.runString(template.source, data, file.pathname)
    return content
  }

  public get fileLoaders() {
    return this.loader.loaders
  }

  public get tokenPreprocessors() {
    return this.env.tokenPreprocessors
  }

  public get utils() {
    return this.env.utils
  }

  public get filters() {
    return this.env.filters
  }

  public get tags() {
    return this.env.tags
  }

  public get cache() {
    return this.env.cache
  }
}
