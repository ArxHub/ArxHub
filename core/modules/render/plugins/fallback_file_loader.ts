import { Plugin } from '~/core/plugin'
import { VirtualFile } from '~/core/modules/vfs/api'
import { RenderEngine } from '../render-engine.ts'
import { TemplateSource } from '~/core/modules/render/api'
import { html, tag } from '@third-party/html'

export class FallbackFileLoaderPlugin implements Plugin<RenderEngine> {
  apply(target: RenderEngine): void {
    target.fileLoaders.push({
      test(_file: VirtualFile): boolean {
        return true
      },

      load(file: VirtualFile): Promise<TemplateSource> {
        const source = html(
          tag('p', [
            'Fallback file loader',
            tag('br'),
            `location: "${file.location}"`,
            tag('br'),
            `kind: "${file.kind}", type: "${file.type}"`,
          ]),
        )

        return Promise.resolve({
          source: source,
          data: {},
        })
      },
    })
  }
}
