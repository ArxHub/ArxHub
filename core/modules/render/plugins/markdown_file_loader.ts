import { VirtualFile } from '~/core/modules/vfs/api'
import { marked } from '@third-party/marked'
import { Plugin } from '~/core/plugin'
import { RenderEngine } from '../render-engine.ts'
import { TemplateSource } from '~/core/modules/render/api'
import { dedent } from '@third-party/dedent'

export class MarkdownFileLoaderPlugin implements Plugin<RenderEngine> {
  apply(target: RenderEngine): void {
    target.filters['__markdown'] = (content: string) => {
      return marked.parse(content, { async: true })
    }

    target.fileLoaders.push({
      test(file: VirtualFile): boolean {
        return file.extension === '.md'
      },

      async load(file: VirtualFile): Promise<TemplateSource> {
        const markdown = await file.text()
        const source = dedent`\
          {{ echo |> await __markdown }}
            ${markdown}
          {{ /echo }}
        `
        return {
          source: source,
        }
      },
    })
  }
}
