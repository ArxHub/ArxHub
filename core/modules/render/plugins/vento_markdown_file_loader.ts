import { VirtualFile } from '~/core/modules/vfs/api'
import { Plugin } from '~/core/plugin'
import { RenderEngine } from '../render-engine.ts'
import { TemplateSource } from '~/core/modules/render/api'
import { dedent } from '@third-party/dedent'

export class VentoMarkdownFileLoaderPlugin implements Plugin<RenderEngine> {
  apply(target: RenderEngine): void {
    target.fileLoaders.push({
      test(file: VirtualFile): boolean {
        return file.extension === '.vtd'
      },

      async load(file: VirtualFile): Promise<TemplateSource> {
        const markdown = await file.text()
        const source = dedent`\
          {{ function __vento }}
            ${markdown}
          {{ /function }}

          {{ __vento() |> await __markdown }}
        `
        return {
          source: source,
        }
      },
    })
  }
}
