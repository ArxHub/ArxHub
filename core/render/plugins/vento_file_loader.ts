import { Plugin } from '~/core/plugin'
import { VirtualFile } from '~/core/vfs/api'
import { RenderEngine } from '~/core/render'
import { TemplateSource } from '~/core/render/api'

export class VentoFileLoaderPlugin implements Plugin<RenderEngine> {
  apply(target: RenderEngine): void {
    target.fileLoaders.push({
      test(file: VirtualFile): boolean {
        return file.extension === '.vto' || file.extension === '.vento'
      },

      async load(file: VirtualFile): Promise<TemplateSource> {
        return {
          source: await file.text(),
          data: {},
        }
      },
    })
  }
}
