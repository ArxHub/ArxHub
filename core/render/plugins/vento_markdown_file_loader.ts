import { Plugin } from '~/core/plugin'
import { dedent } from '@third-party/dedent'
import { RenderEngine } from '~/core/render'
import { VirtualFile } from '~/core/vfs/api'
import { TemplateSource } from '~/core/render/api'

export class VentoMarkdownFileLoaderPlugin implements Plugin<RenderEngine> {
  apply(target: RenderEngine): void {
    target.fileLoaders.push({
      test(file: VirtualFile): boolean {
        return file.extension === '.vtd'
      },

      async load(file: VirtualFile): Promise<TemplateSource> {
        const markdown = await file.text()
        const source = dedent`\
          {{ function __vento }}
            ${markdown}
          {{ /function }}

          {{ __vento() |> await __markdown }}
        `
        return {
          source: source,
        }
      },
    })
  }
}
