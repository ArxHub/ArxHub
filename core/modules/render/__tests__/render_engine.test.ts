import { afterEach, describe, it } from 'jsr:@std/testing/bdd'
import { createAssertSnapshot } from 'jsr:@std/testing/snapshot'
import { RenderEngine } from '../render-engine.ts'
import { LocalFileSystem } from '../../../vfs/local-file-system/mod.ts'
import { FallbackFileLoaderPlugin } from '../plugins/fallback_file_loader.ts'
import { VentoFileLoaderPlugin } from '../plugins/vento_file_loader.ts'
import { MarkdownFileLoaderPlugin } from '../plugins/markdown_file_loader.ts'
import { VentoMarkdownFileLoaderPlugin } from '../plugins/vento_markdown_file_loader.ts'

const assertSnapshot = createAssertSnapshot({
  dir: `${import.meta.dirname}/snapshots`,
})

describe('render engine', () => {
  const vfs = new LocalFileSystem('local', `${import.meta.dirname}/wikidata`)
  const engine = new RenderEngine(vfs, [
    new VentoMarkdownFileLoaderPlugin(),
    new MarkdownFileLoaderPlugin(),
    new VentoFileLoaderPlugin(),
    new FallbackFileLoaderPlugin(),
  ])

  afterEach(() => {
    engine.cache.clear()
  })

  it('fallback loader', async (t) => {
    const file = await vfs.file('fallback.ext')
    const content = await engine.render(file, { mode: 'static' })
    await assertSnapshot(t, content)
  })

  it('vento', async (t) => {
    const file = await vfs.file('vento.vto')
    const content = await engine.render(file, { mode: 'static', data: { foo: 'foo', bar: 'bar' } })
    await assertSnapshot(t, content)
  })

  it('markdown', async (t) => {
    const file = await vfs.file('markdown.md')
    const content = await engine.render(file, { mode: 'static', data: { foo: 'foo', bar: 'bar' } })
    await assertSnapshot(t, content)
  })

  it('vento and markdown', async (t) => {
    const file = await vfs.file('vento_and_markdown.vtd')
    const content = await engine.render(file, { mode: 'static', data: { foo: 'foo', bar: 'bar' } })
    await assertSnapshot(t, content)
  })

  it('include works', async (t) => {
    const file = await vfs.file('include.vto')
    const content = await engine.render(file, { mode: 'static', data: { foo: 'foo', bar: 'bar' } })
    await assertSnapshot(t, content)
  })
})
