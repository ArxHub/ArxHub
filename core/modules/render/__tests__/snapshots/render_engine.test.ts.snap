export const snapshot = {};

snapshot[`render engine > fallback loader 1`] = `
"<p>
  Fallback file loader
  <br>
  location: &quot;local:fallback.ext&quot;
  <br>
  kind: &quot;unknown&quot;, type: &quot;application/octet-stream&quot;
</p>
"
`;

snapshot[`render engine > vento 1`] = `
'Simple template

Foo: "foo"
Bar: "bar"
Foo + Bar: "foobar"
'
`;

snapshot[`render engine > markdown 1`] = `
'<h1>Headers</h1>
<h1>h1 Heading</h1>
<h2>h2 Heading</h2>
<h3>h3 Heading</h3>
<h4>h4 Heading</h4>
<h5>h5 Heading</h5>
<h6>h6 Heading</h6>
<p>Alternatively, for H1 and H2, an underline-ish style:</p>
<h1>Alt-H1</h1>
<h2>Alt-H2</h2>
<hr>
<h1>Emphasis</h1>
<p>Emphasis, aka italics, with <em>asterisks</em> or <em>underscores</em>.</p>
<p>Strong emphasis, aka bold, with <strong>asterisks</strong> or <strong>underscores</strong>.</p>
<p>Combined emphasis with <strong>asterisks and <em>underscores</em></strong>.</p>
<p>Strikethrough uses two tildes. <del>Scratch this.</del></p>
<p><strong>This is bold text</strong></p>
<p><strong>This is bold text</strong></p>
<p><em>This is italic text</em></p>
<p><em>This is italic text</em></p>
<p><del>Strikethrough</del></p>
<hr>
<h1>Lists</h1>
<ol>
<li>First ordered list item</li>
<li>Another item</li>
</ol>
<ul>
<li>Unordered sub-list.</li>
</ul>
<ol>
<li>Actual numbers don&#39;t matter, just that it&#39;s a number</li>
<li>Ordered sub-list</li>
<li>And another item.</li>
</ol>
<p>You can have properly indented paragraphs within list items. Notice the blank line above, and the leading spaces (at least one, but we&#39;ll use three here to also align the raw Markdown).</p>
<p> To have a line break without a paragraph, you will need to use two trailing spaces.<br>   Note that this line is separate, but within the same paragraph.<br>   (This is contrary to the typical GFM line break behaviour, where trailing spaces are not required.)</p>
<ul>
<li>Unordered list can use asterisks</li>
</ul>
<ul>
<li>Or minuses</li>
</ul>
<ul>
<li>Or pluses</li>
</ul>
<ol>
<li>Make my changes<ol>
<li>Fix bug</li>
<li>Improve formatting<ul>
<li>Make the headings bigger</li>
</ul>
</li>
</ol>
</li>
<li>Push my commits to GitHub</li>
<li>Open a pull request<ul>
<li>Describe my changes</li>
<li>Mention all the members of my team<ul>
<li>Ask for feedback</li>
</ul>
</li>
</ul>
</li>
</ol>
<ul>
<li>Create a list by starting a line with <code>+</code>, <code>-</code>, or <code>*</code></li>
<li>Sub-lists are made by indenting 2 spaces:<ul>
<li>Marker character change forces new list start:<ul>
<li>Ac tristique libero volutpat at</li>
</ul>
<ul>
<li>Facilisis in pretium nisl aliquet</li>
</ul>
<ul>
<li>Nulla volutpat aliquam velit</li>
</ul>
</li>
</ul>
</li>
<li>Very easy!</li>
</ul>
<hr>
<h1>Task lists</h1>
<ul>
<li><input checked="" disabled="" type="checkbox"> Finish my changes</li>
<li><input disabled="" type="checkbox"> Push my commits to GitHub</li>
<li><input disabled="" type="checkbox"> Open a pull request</li>
<li><input checked="" disabled="" type="checkbox"> @mentions, #refs, <a href="">links</a>, <strong>formatting</strong>, and <del>tags</del> supported</li>
<li><input checked="" disabled="" type="checkbox"> list syntax required (any unordered or ordered list supported)</li>
<li><input disabled="" type="checkbox"> this is a complete item</li>
<li><input disabled="" type="checkbox"> this is an incomplete item</li>
</ul>
<hr>
<p>Foo: &quot;{{ foo }}&quot;
Bar: &quot;{{ bar }}&quot;
Foo + Bar: &quot;{{ foo + bar }}&quot;</p>
<hr>

'
`;

snapshot[`render engine > vento and markdown 1`] = `
"

<h1>H1 Header</h1>
<ul>
<li>Foo: &quot;foo&quot;</li>
<li>Bar: &quot;bar&quot;</li>
<li>Foo + Bar: &quot;foobar&quot;</li>
</ul>
<hr>
<h2>H2 Header</h2>

"
`;
